pkg/reconciler/ingress/resources/httproute.go

```diff
@@ -0,0 +1,158 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resources
+
+import (
+	"context"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/utils/pointer"
+	gwv1alpha1 "sigs.k8s.io/gateway-api/apis/v1alpha1"
+
+	"knative.dev/networking/pkg"
+	netv1alpha1 "knative.dev/networking/pkg/apis/networking/v1alpha1"
+	"knative.dev/pkg/kmeta"
+
+	"knative.dev/net-gateway-api/pkg/reconciler/ingress/config"
+)
+
+// MakeHTTPRoute creates HTTPRoute to set up routing rules.
+func MakeHTTPRoute(
+	ctx context.Context,
+	ing *netv1alpha1.Ingress,
+	rule *netv1alpha1.IngressRule,
+) (*gwv1alpha1.HTTPRoute, error) {
+
+	visibility := ""
+	if rule.Visibility == netv1alpha1.IngressVisibilityClusterLocal {
+		visibility = "cluster-local"
+	}
+
+	return &gwv1alpha1.HTTPRoute{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      LongestHost(rule.Hosts),
+			Namespace: ing.Namespace,
+			Labels: kmeta.UnionMaps(ing.Labels, map[string]string{
+				pkg.VisibilityLabelKey: visibility,
+			}),
+			Annotations: kmeta.FilterMap(ing.GetAnnotations(), func(key string) bool {
+				return key == corev1.LastAppliedConfigAnnotation
+			}),
+			OwnerReferences: []metav1.OwnerReference{*kmeta.NewControllerRef(ing)},
+		},
+		Spec: makeHTTPRouteSpec(ctx, rule),
+	}, nil
+}
+
+func makeHTTPRouteSpec(
+	ctx context.Context,
+	rule *netv1alpha1.IngressRule,
+) gwv1alpha1.HTTPRouteSpec {
+
+	hostnames := []gwv1alpha1.Hostname{}
+	for _, hostname := range rule.Hosts {
+		hostnames = append(hostnames, gwv1alpha1.Hostname(hostname))
+	}
+
+	rules := makeHTTPRouteRule(rule)
+
+	gatewayConfig := config.FromContext(ctx).Gateway
+	ns, name, _ := cache.SplitMetaNamespaceKey(gatewayConfig.LookupGateway(rule.Visibility))
+
+	gatewayRef := gwv1alpha1.GatewayReference{
+		Namespace: ns,
+		Name:      name,
+	}
+
+	return gwv1alpha1.HTTPRouteSpec{
+		Hostnames: hostnames,
+		Rules:     rules,
+		Gateways: &gwv1alpha1.RouteGateways{
+			Allow:       gatewayAllowTypePtr(gwv1alpha1.GatewayAllowFromList),
+			GatewayRefs: []gwv1alpha1.GatewayReference{gatewayRef},
+		},
+	}
+}
+
+func makeHTTPRouteRule(rule *netv1alpha1.IngressRule) []gwv1alpha1.HTTPRouteRule {
+	rules := []gwv1alpha1.HTTPRouteRule{}
+
+	for _, path := range rule.HTTP.Paths {
+		var forwards []gwv1alpha1.HTTPRouteForwardTo
+		var preFilters []gwv1alpha1.HTTPRouteFilter
+
+		if path.AppendHeaders != nil {
+			preFilters = []gwv1alpha1.HTTPRouteFilter{{
+				Type: gwv1alpha1.HTTPRouteFilterRequestHeaderModifier,
+				RequestHeaderModifier: &gwv1alpha1.HTTPRequestHeaderFilter{
+					Set: path.AppendHeaders,
+				}}}
+		}
+
+		set := map[string]string{}
```

Comment:
nit:

```suggestion
		var set map[string]string
```

avoids an unnecessary allocation here.